source/lac/trilinos_precondition_ml.cc

```diff
@@ -183,32 +183,38 @@ namespace TrilinosWrappers
     if (constant_modes_dimension > 0)
       {
         const size_type n_rows = n_global_rows(matrix);
-        const bool constant_modes_are_global =
-          additional_data.constant_modes[0].size() == n_rows;
-        const size_type n_relevant_rows =
-          constant_modes_are_global ? n_rows : additional_data.constant_modes[0].size();
+        const bool constant_modes_are_global
+          = additional_data.constant_modes[0].size() == n_rows;
         const size_type my_size = domain_map.NumMyElements();
-        if (constant_modes_are_global == false)
-          Assert (n_relevant_rows == my_size,
-                  ExcDimensionMismatch(n_relevant_rows, my_size));
         Assert (n_rows ==
                 static_cast<size_type>(global_length(distributed_constant_modes)),
                 ExcDimensionMismatch(n_rows,
                                      global_length(distributed_constant_modes)));
 
-        (void)n_relevant_rows;
-        (void)global_length;
-
         // Reshape null space as a contiguous vector of doubles so that
         // Trilinos can read from it.
-        for (size_type d=0; d<constant_modes_dimension; ++d)
-          for (size_type row=0; row<my_size; ++row)
-            {
-              TrilinosWrappers::types::int_type global_row_id =
-                constant_modes_are_global ? gid(domain_map,row) : row;
-              distributed_constant_modes[d][row] =
-                additional_data.constant_modes[d][global_row_id];
-            }
+        if (constant_modes_are_global)
```

Comment:
I don't know if you will gain anything by duplicating code. ```constant_modes_are_global``` is  constant so the compiler should be able to write the same code you wrote. Worst-case scenario, the branch-predictor should nailed it. I am not sure if it is worth duplicate code.